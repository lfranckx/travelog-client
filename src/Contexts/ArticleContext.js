/*eslint semi: ["error", "always"]*/
import React, { Component } from 'react';

export const nullArticle = {
    author: {},
    tags: [],
};
export const nullAuthor = {};
export const nullUser = {};
export const nullUsersArticles = [];

const ArticleContext = React.createContext({
    error: null,
    article: nullArticle,
    author: nullAuthor,
    user: nullUser,
    articlesList: [],
    authorsList: [],
    usersArticles: nullUsersArticles,
    setError: () => {},
    clearError: () => {},
    setArticle: () => {},
    clearArticle: () => {},
    setAuthor: () => {},
    clearAuthor: () => {},
    setUser: () => {},
    clearUser: () => {},
    setArticlesList: () => {},
    setAuthorsList: () => {},
    setUsersArticles: () => {},
    clearUsersArticles: () => {}
});
export default ArticleContext;

export class ArticleProvider extends Component {
    state = {
        error: null,
        article: nullArticle,
        author: nullAuthor,
        user: nullUser,
        articlesList: [],
        authorsList: [],
        usersArticles: []
    };

    setError = error => {
        console.error(error);
        this.setState({ error });
    };

    clearError = () => {
        this.setState({ error: null });
    };

    setArticle = article => {
        this.setState({ article });
    };

    clearArticle = () => {
        this.setArticle(nullArticle);
    };

    setAuthor = author => {
        this.setState({ author });
    };

    clearAuthor = () => {
        this.setAuthor(nullAuthor);
    };

    setUser = user => {
        console.log('setting user:', user);
        this.setState({ user });
    };

    clearUser = () => {
        this.setState(nullUser);
    };

    setArticlesList = articlesList => {
        this.setState({ articlesList });
    };

    setAuthorsList = authorsList => {
        this.setState({ authorsList });
    };

    setUsersArticles = usersArticles => {
        this.setState({ usersArticles });
    };

    clearUsersArticles = () => {
        this.setUsersArticles(nullUsersArticles);
    };

    render() {
        const value = {
            error: this.state.error,
            article: this.state.article,
            author: this.state.author,
            user: this.state.user,
            articlesList: this.state.articlesList,
            authorsList: this.state.authorsList,
            usersArticles: this.state.usersArticles,
            setError: this.setError,
            clearError: this.clearError,
            setArticle: this.setArticle,
            clearArticle: this.clearArticle,
            setAuthor: this.setAuthor,
            clearAuthor: this.clearAuthor,
            setUser: this.setUser,
            clearUser: this.clearUser,
            setArticlesList: this.setArticlesList,
            setAuthorsList: this.setAuthorsList,
            setUsersArticles: this.setUsersArticles,
            clearUsersArticles: this.clearUsersArticles
        };
        
        return (
            <ArticleContext.Provider value={value}>
                {this.props.children}
            </ArticleContext.Provider>
        );
    }   
}