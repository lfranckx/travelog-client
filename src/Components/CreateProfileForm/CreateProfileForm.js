import React, { Component } from 'react';
import ArticleContext from '../../Contexts/ArticleContext';
import AuthorApiService from '../../Services/author-api-service';

export default class CreateProfileForm extends Component {

    static contextType = ArticleContext;

    constructor(props) {
        super(props);
        this.fileInput = React.createRef();
        this.state = {
            error: null,
            fileSelected: null
        };
    }

    handleCreateProfile = ev => {
        ev.preventDefault();
        this.setState({ error: null });
        const { name, about } = ev.target;
        const { user } = this.context;
        const author = {
            username: user.username,
            name: name.value,
            about: about.value
        }
        const fileSelected = this.fileInput.current.files[0];
        let data = new FormData();
        data.append('image', fileSelected);

        AuthorApiService.updateAuthor(author)
            .then(this.context.setUser(author))
            .then(res => {
                if (fileSelected) {
                    AuthorApiService.uploadFile(data)
                        .then(res => {
                            (!res.ok) ? res.json().then(e => Promise.reject(e))
                                : res.json()
                            .then(data => {
                                author.profile_image = data.image_url;
                                AuthorApiService.updateAuthor(author)
                                .then(this.context.setUser(author))
                                .then(this.props.onSubmitForm)
                                .catch(this.context.setError);
                            })
                        });
                }
            })
            .then(this.props.onSubmitForm)
            .catch(this.context.setError);

        // if (fileSelected) {
        //     AuthorApiService.uploadFile(data)
        //         .then(res => {
        //             (!res.ok) ? res.json().then(e => Promise.reject(e))
        //                 : res.json()
        //             .then(data => {
        //                 user.profile_image = data.image_url;
        //                 AuthorApiService.updateAuthor(user)
        //                 .then(this.context.setUser(user))
        //                 .then(this.props.onSubmitForm)
        //                 .catch(this.context.setError);
        //             })
        //         });
        // } else if (!fileSelected) {
            // AuthorApiService.updateAuthor(author)
            //     .then(this.context.setUser(author))
                // .then(this.props.onSubmitForm)
                // .catch(this.context.setError);
        // }
    }

    render() {
        const { error } = this.state;
        const { user } = this.context;
        return (
            <form className="edit-profile-form" onSubmit={this.handleCreateProfile}>
                <div role='alert'>{error && <p className='error'>{error}</p>}</div>
                <div>
                    <label>Select a profile image</label>
                </div>
                <div>
                    <input
                        ref={this.fileInput}
                        type="file"
                        accept=".png, .jpg, .jpeg .gif"
                        name="file" 
                        aria-label='file'
                        className="file"
                    />
                </div>
                <div>
                    <label>Full Name</label>
                </div>
                <div>
                    <input className='name'
                        type='text'
                        name='name'
                        aria-label='name'
                        required
                        defaultValue={user.name}
                    />
                </div>
                <div>
                    <label>About</label>
                </div>
                <div>
                    <textarea
                        rows='18'
                        name='about'
                        aria-label='about'
                    />
                </div>
                <div className="form-button">
                    <button type='submit'>Submit</button>
                </div>
            </form>
        );
    }
}