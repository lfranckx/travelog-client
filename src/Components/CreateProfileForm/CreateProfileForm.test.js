import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import CreateProfileForm from './CreateProfileForm';
import { ArticleProvider } from '../../Contexts/ArticleContext';

describe(`CreateProfileForm`, () => {
    const props = {
        handleSubmitForm: () => {}
    };

    it(`renders form given props`, () => {
        const wrapper = shallow(
            <ArticleProvider>
                <CreateProfileForm {...props}/>
            </ArticleProvider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});