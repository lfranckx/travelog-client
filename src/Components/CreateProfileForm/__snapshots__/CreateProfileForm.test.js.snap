// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProfileForm renders form given props 1`] = `
<ContextProvider
  value={
    Object {
      "active": false,
      "addComment": [Function],
      "article": Object {},
      "articlesList": Array [],
      "author": Object {},
      "authorsList": Array [],
      "clearArticle": [Function],
      "clearAuthor": [Function],
      "clearError": [Function],
      "clearUser": [Function],
      "clearUsersArticles": [Function],
      "comments": Array [],
      "error": null,
      "search": "",
      "setArticle": [Function],
      "setArticlesList": [Function],
      "setAuthor": [Function],
      "setAuthorsList": [Function],
      "setComments": [Function],
      "setError": [Function],
      "setUser": [Function],
      "setUsersArticles": [Function],
      "toggleActive": [Function],
      "updateSearch": [Function],
      "user": Object {},
      "usersArticles": Array [],
    }
  }
>
  <CreateProfileForm
    handleSubmitForm={[Function]}
  />
</ContextProvider>
`;
